### ISSUE TITLE
piHole updateGravity "DNS resolution is currently unavailable"

### ISSUE THREAD

**Issue by sm1ddy** [pi-hole/pi-hole#1801]
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

`{1}`

---
**Expected behaviour:**

`{I run PiHole on a Synology NAS in Debian Chroot. When running piHole updateGravity from the admin panel or terminal I expect the update to complete.}`

**Actual behaviour:**

`{Running piHole updateGravity in the admin panel or terminal returns the error "DNS resolution is currently unavailable"}`

**Steps to reproduce:**

`{Uncertain, maybe try to use a Synology NAS environment or something running Debian Chroot}`

**Debug token provided by [uploading `pihole -d` log](https://discourse.pi-hole.net/t/the-pihole-command-with-examples/738#debug):**

`{4xeg3jnc5w}`

**Troubleshooting undertaken, and/or other relevant information:**

`{I've tried multiple reinstalls and rolling back to previous master tags. Still seems to be broken}`

**Comment by dschaper** [comment]
Thank you for the report, we are looking in to this. None of the developers have Synology devices so we may not be able to duplicate it.

**Comment by Failure404** [comment]
Same happens to me on a RPI2 and RPI3.
I commented out the dns check in gravity.sh, it updates fine after that. So DNS resolution is working, its just the check that fails.

**Comment by dschaper** [comment]
`getent hosts pi.hole` is returning nothing?

**Comment by Failure404** [comment]
Affirmative. Nothing is returned.

**Comment by dschaper** [comment]
Thank you for the report. We'll take a look.

**Comment by dschaper** [comment]
Can either (or both) of you post the contents of your `/etc/resolv.conf` file?

**Comment by sm1ddy** [comment]
Hi @dschaper, thanks for the quick response mate.

My resolv.conf file contains the following line (the ip being my router).
nameserver      192.168.1.254

**Comment by sm1ddy** [comment]
Awesome, thanks for the great project btw!

**Comment by natemccurdy** [comment]
Changing `/etc/resolv.conf` to point to 127.0.0.1 worked for me.

```
nameserver 127.0.0.1
```

**Comment by dschaper** [comment]
https://github.com/pi-hole/pi-hole/issues/1806

**Comment by sschadwick** [comment]
Also happening on Digital Ocean:
```
  [i] Running gravity.sh
  [✗] DNS resolution is currently unavailable
  [✗] DNS resolution is not available
Unable to complete update, contact Pi-hole
```

**Comment by sschadwick** [comment]
What are the steps to recover from this? I'm still able to access my admin interface, however I no longer have connection statistics at the top, the FTL graphs fail to load and spin endlessly, and the FTL version is listed as `vDev-437af07` (I'm not on dev branch to my knowledge). I would love to resolve this without having to reinstall pihole entirely.

**Comment by dschaper** [comment]
`vDev-437af07` is the latest released version of FTL, so you are okay there. For a temporary workaround you can edit `/etc/resolv.conf` and comment out any `nameserver` lines that are not `127.0.0.1` and then try again. That should get you past the `DNS resolution` problems. If you still have problems with the web interface you can try `pihole -r` and repair and make sure you flush the browser cache on the client machine.

**Comment by sschadwick** [comment]
@dschaper That fixed it! Looks like I had a corrupt download of FTL in addition to the DNS resolution bug. The repair command fixed both (once I had removed the additional resolvers). Does this mean that only 1/3 of my DNS traffic was being routed through Pi-hole?

Here's the doc I used to change the DNS on my DigitalOcean droplet: https://www.digitalocean.com/community/questions/how-do-i-switch-my-dns-resolvers-away-from-google

**Comment by dschaper** [comment]
The `/etc/resolv.conf` file is the list of nameservers that the droplet itself uses to resolve names to IPs. The `127.0.0.1` entry tells the droplet to use the Pi-hole `dnsmasq` program directly, so the Pi-hole droplet is using the Pi-hole filtering itself. When all 3 nameservers were listed the droplet wasn't filtering queries, but Pi-hole was still working and all clients that were set to use the droplets IP address as the DNS server were getting the full use of Pi-hole. 

In short, with all 3 nameservers you could `curl` a blocked site when you were at the command line of the droplet as the droplet wasn't being filtered. But that only affects the droplet directly and commands done at the terminal. All clients were always filtered.

**Comment by jaspov** [comment]
same issue with dns resolution on my oi. Checked resolv.conf and it has only:
nameserver 127.0.0.1 

any other way to fix this?

**Comment by Romijn** [comment]
I have the same issue, putting # in front of the nameservers solved the update issue, but should I then remove the # again? After pi-hole is updated and started?

**Comment by vr6syncro** [comment]
same here

**Comment by mcfalld** [comment]
the same issue with DNS resolution here. my etc/resolv.conf also only has 127.0.0.1 but mine is also not working. debug token is: huw6qn7kz2 not trying to jack the thread just don't want to start a new one for the same issue. My entire network does not have internet access and network troubleshooting says DNS server not responding. Had to change my DNS servers on my router back to "auto assign" to restore network access to my home.

**Comment by Mcat12** [comment]
@mcfalld There's some odd entries in your gravity.list file, specifically the first and last in the file: `>` and `Binary file /etc/pihole/list.preEventHorizon matches`.
Try removing those.

**Comment by mcfalld** [comment]
I deleted those and rebooted but still "DNS server isnt responding." And Dns resolution is currently unavailable when i do the pihole -r repair option. It waits for so many seconds then it says dns resolution is now available. but network connectivity is not restored.

**Comment by Mcat12** [comment]
What is the output of
```
dig pi.hole @127.0.0.1
```

**Comment by mcfalld** [comment]
while my dns servers are my pihole ip address : 
pi@raspberrypi:~ $ dig pi.hole @127.0.0.1

; <<>> DiG 9.10.3-P4-Raspbian <<>> pi.hole @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 6144
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;pi.hole.                       IN      A

;; ANSWER SECTION:
pi.hole.                2       IN      A       192.168.1.200

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Fri Dec 08 20:49:54 CST 2017
;; MSG SIZE  rcvd: 52

while my dns servers are automatically assigned : 

pi@raspberrypi:~ $ dig pi.hole @127.0.0.1

; <<>> DiG 9.10.3-P4-Raspbian <<>> pi.hole @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 13283
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;pi.hole.                       IN      A

;; ANSWER SECTION:
pi.hole.                2       IN      A       192.168.1.200

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Fri Dec 08 20:59:33 CST 2017
;; MSG SIZE  rcvd: 52

**Comment by p1r473** [comment]
I think I'm having the same issue. DNS resolution on my Pi is busted
nameserver 127.0.0.1 doesn't work
nameserver 8.8.8.8 works

To answer @Mcat12 question:

i@RaspberryPi:~ $ dig pi.hole @127.0.0.1

; <<>> DiG 9.10.3-P4-Raspbian <<>> pi.hole @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 24824
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;pi.hole.			IN	A

;; ANSWER SECTION:
pi.hole.		2	IN	A	127.0.0.1

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Sat Dec 09 04:06:32 UTC 2017
;; MSG SIZE  rcvd: 52

**Comment by PermanentWaves** [comment]
Same issue here. The DNS was working fine but the check was failing. Adding pi.hole to /etc/hosts allowed the check to pass.

**Comment by wnoisephx** [comment]
I had the same issue, but solved it by adding the line
127.0.0.1 pi.hole
to my /etc/hosts file

**Comment by Mcat12** [comment]
@mcfalld You might be limited by the GitHub API. It only lets you access it about 60 times per hour. What is the output of `getent hosts pi.hole`?

@p1r473 What is the output of the above command? Also, run `pihole -d` for a debug token.

**Comment by mcfalld** [comment]
@Mcat12  I appreciate you taking the time to help. 
pi@raspberrypi:~ $ getent hosts pi.hole
192.168.1.200   pi.hole

what all counts for accessing the API? I cant see why I would be accessing it more than 60 times per hour.

**Comment by p1r473** [comment]
@Mcat12 

getent hosts pi.hole:
127.0.0.1 pi.hole

debug token is: act037ymzm

I snapped a pic on boot of my pi- one line mentions DNS
https://imgur.com/m0iVamV

Yesterday, I kept having to change /etc/resolv.conf from 127.0.0.1 to 8.8.8.8 because I couldn't even ping Google from the Pi.

**Comment by Mcat12** [comment]
@mcfalld The background check for the latest version which happens every 10min and checking the versions of core, web, and FTL as part of the upgrade/repair scripts (in your case, only FTL would be applicable). You can check this by running `curl https://api.github.com/rate_limit`. As long as you aren't rate limited and the getent command is successful it should work... If it still does not, then what is the output of this command?
```
curl -sI https://github.com/pi-hole/FTL/releases/latest
```

@p1r473 it looks like your IPv4 address is set to `127.0.0.1`. Use `pihole -r` to change it to the IP address of the Pi-hole device.

**Comment by p1r473** [comment]
@Mcat12 thanks, seems to have fixed it for now anyways! Ill post back if anything changes

**Comment by mcfalld** [comment]
pi@raspberrypi:~ $ curl  https://api.github.com/rate_limit
{
  "resources": {
    "core": {
      "limit": 60,
      "remaining": 54,
      "reset": 1512885002
    },
    "search": {
      "limit": 10,
      "remaining": 10,
      "reset": 1512882146
    },
    "graphql": {
      "limit": 0,
      "remaining": 0,
      "reset": 1512885686
    }
  },
  "rate": {
    "limit": 60,
    "remaining": 54,
    "reset": 1512885002
  }
}

pi@raspberrypi:~ $ curl -sI https://github.com/pi-hole/FTL/releases/latest
HTTP/1.1 302 Found
Server: GitHub.com
Date: Sun, 10 Dec 2017 05:02:23 GMT
Content-Type: text/html; charset=utf-8
Status: 302 Found
Cache-Control: no-cache
Vary: X-PJAX
Location: https://github.com/pi-hole/FTL/releases/tag/v2.12
X-UA-Compatible: IE=Edge,chrome=1
Set-Cookie: logged_in=no; domain=.github.com; path=/; expires=Thu, 10 Dec 2037 05:02:23 -0000; secure; HttpOnly
Set-Cookie: _gh_sess=eyJzZXNzaW9uX2lkIjoiZGYxNjNkZmEwNjJjNTQ1ZWI3MDg1Yjc2YjcyMjIyNzIiLCJsYXN0X3JlYWRfZnJvbV9yZXBsaWNhcyI6MTUxMjg4MjE0Mjk5OCwic3B5X3JlcG8iOiJwaS1ob2xlL0ZUTCIsInNweV9yZXBvX2F0IjoxNTEyODgyMTQzfQ%3D%3D--84134e6b8a3b0d080f2c639d4aa678faa073040c; path=/; secure; HttpOnly
X-Request-Id: 14b3cc639081f36b6ac1af37fe9a8d54
X-Runtime: 0.058866
Expect-CT: max-age=2592000, report-uri="https://api.github.com/_private/browser/errors"
Content-Security-Policy: default-src 'none'; base-uri 'self'; block-all-mixed-content; child-src render.githubusercontent.com; connect-src 'self' uploads.github.com status.github.com collector.githubapp.com api.github.com www.google-analytics.com github-cloud.s3.amazonaws.com github-production-repository-file-5c1aeb.s3.amazonaws.com github-production-upload-manifest-file-7fdce7.s3.amazonaws.com github-production-user-asset-6210df.s3.amazonaws.com wss://live.github.com; font-src assets-cdn.github.com; form-action 'self' github.com gist.github.com; frame-ancestors 'none'; img-src 'self' data: assets-cdn.github.com identicons.github.com collector.githubapp.com github-cloud.s3.amazonaws.com *.githubusercontent.com; media-src 'none'; script-src assets-cdn.github.com; style-src 'unsafe-inline' assets-cdn.github.com
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
Public-Key-Pins: max-age=0; pin-sha256="WoiWRyIOVNa9ihaBciRSC7XHjliYS9VwUGOIud4PB18="; pin-sha256="RRM1dGqnDFsCJXBTHky16vi1obOlCgFFn/yOhI/y+ho="; pin-sha256="k2v657xBsOVe1PQRwOsHsw3bsGT2VzIqz5K+59sNQws="; pin-sha256="K87oWBWM9UZfyddvDfoxL+8lpNyoUB2ptGtn0fv6G2Q="; pin-sha256="IQBnNBEiFuhj+8x6X8XLgh01V9Ic5/V3IRQLNFFc7v4="; pin-sha256="iie1VXtL7HzAMF+/PVPR9xzT80kQxdZeJ+zduCB3uj0="; pin-sha256="LvRiGEjRqfzurezaWuj8Wie2gyHMrW5Q06LspMnox7A="; includeSubDomains
X-Content-Type-Options: nosniff
X-Frame-Options: deny
X-XSS-Protection: 1; mode=block
X-Runtime-rack: 0.067715
X-GitHub-Request-Id: 8AC6:7AC1:3EBC1A8:6ACB986:5A2CBFDE

It is still not working. I have repaired, and reconfigured. I still currently have my dns servers assigned automatically because when I set it to 192.168.1.200 (my pi ip address) It still says it has problems with dns resolution. Im hoping you see some error in the results from the commands above.
Thank you

**Comment by carlitem** [comment]
I have PiHole on my Synology, just run (in chroot) `apt-get update`, then `apt-get upgrade`... lots of packages where upgraded! After that to check, run `pihole -g` works for me without any file modification!

**Comment by PromoFaux** [comment]
@JamborJan you could try switching to `development` by running `pihole checkout dev`, as it currently has the fixes in. Fingers crossed we'll be releasing 3.2.1 very soon[™](https://wow.gamepedia.com/Soon)

Once we release, you can switch back to master with `pihole checkout master`

**Comment by JamborJan** [comment]
Thanks @PromoFaux that did work. Will go back as soon as the release is available.

**Comment by Mcat12** [comment]
@mcfalld What is the output of these commands?
```
latesttag=$(curl -sI https://github.com/pi-hole/FTL/releases/latest | grep "Location" | awk -F '/' '{print $NF}')
echo $latesttag
if [[ ! "${latesttag}" == v* ]]; then echo true; else echo false; fi
```

**Comment by mcfalld** [comment]
pi@raspberrypi:~ $ curl -sI https://github.com/pi-hole/FTL/releases/latest | grep "Location" | awk -F '/' '{print $NF}'
v2.12

pi@raspberrypi:~ $ echo $latesttag

pi@raspberrypi:~ $ if [[ ! "${latesttag}" == v* ]]; then echo true; else echo false; fi
true
pi@raspberrypi:~ $

**Comment by Mcat12** [comment]
You missed the `latesttag=` on the first command

**Comment by mcfalld** [comment]
pi@raspberrypi:~ $ latesttag=$(curl -sI https://github.com/pi-hole/FTL/releases/latest | grep "Location" | awk -F '/' '{print $NF}')
pi@raspberrypi:~ $ echo $latesttag
v2.12
pi@raspberrypi:~ $ if [[ ! "${latesttag}" == v* ]]; then echo true; else echo false; fi
false

**Comment by Mcat12** [comment]
Hm, try running the reconfigure again? Maybe dnsmasq hasn't started....

**Comment by mcfalld** [comment]
curl: (6) Could not resolve host: github.com
  [i] Corruption detected...
  [✗] Downloading and Installing FTL
  Error: Unable to get latest release location from GitHub
  [✗] FTL Engine not installed

**Comment by mcfalld** [comment]
Strange the blocking appears to be working over the openVPN connection. Im not at home so I am not sure if the blocking is working while on the network. As soon as I get home I will change the settings for my router DNS servers to use my pi and see if the DNS resolution is still an issue.

**Comment by Mcat12** [comment]
What interface is dnsmasq listening on? Check in the settings page. You may need to tell it to listen on more interfaces, since it seems to only be working on certain connections.

**Comment by mcfalld** [comment]
it is listening on eth1 which is a usb to ethernet adapter for a little more speed.

**Comment by Mcat12** [comment]
Run `pihole -d` for a debug token. Maybe dnsmasq is failing before the reconfigure and is only started when the script finishes...

**Comment by mcfalld** [comment]
7gyfltai9e is my token. sorry i was away for the weekend.

**Comment by mcfalld** [comment]
After an uninstall and attempted resinstall I am now getting this message 

  [i] Logging On.
  [✗] Check for existing repository in /etc/.pihole
  [i] Clone https://github.com/pi-hole/pi-hole.git into /etc/.pihole...
  Error: Could not update local repository. Contact support.

**Comment by Mcat12** [comment]
Are you running an OctoPi install? If not, then what happens if you run `sudo git clone https://github.com/pi-hole/pi-hole.git /etc/.pihole`?

**Comment by dschaper** [comment]
Closed via #1856
